machine/machine_core/ssh_connection.py

```diff
@@ -417,16 +417,23 @@ def write(self, dest, content, append=False, owner=None, perm=None):
             append: If True, append to existing file instead of replacing it
             owner: If set, call chown on the file with the given owner string
             perm: Optional file permission as chmod shell string (e.g. "0600")
+
+        The directory of dest is created automatically.
         """
         assert dest
         assert self.ssh_address
 
-        cmd = "cat %s '%s'" % (append and '>>' or '>', dest)
+        qdest = shlex.quote(dest)
+
+        cmds = [
+            f"mkdir -p {shlex.quote(os.path.dirname(dest))}",
+            f"cat {append and '>>' or '>'} {qdest}",
+        ]
         if owner:
-            cmd += " && chown '%s' '%s'" % (owner, dest)
+            cmds.append(f"chown {owner} {qdest}")
         if perm:
-            cmd += " && chmod '%s' '%s'" % (perm, dest)
-        self.execute(cmd, input=content)
+            cmds.append(f"chmod {perm} {qdest}")
+        self.execute('\n'.join(cmds), input=content)
```

Comment:
You couldn't resist, huh? :)

I'd have preferred to see like

```python
cmd = """
  foo
  bar
"""

cmd+="baz\n"
cmd+="quux\n"
```

but this is fine too :)

Reply:
It's almost like it would be useful to have a variant of `m.execute()` that accepted multiple commands... ;)

Reply:
OR: would it really be so evil to just use separate `m.execute()` invocations?  Keep in mind that we use the SSH control channel mechanism.  It's probably lower overhead than we think...

Reply:
I had that as well, it just looks more ugly in `-tv` output. Also this mixes multi-line with single-line and `\n`, so worst of both worlds. I have no strong opinion either way though, if you prefer that I can do that.

Reply:
Right, separate execute() are fine. Reworked accordingly, PTAL?